core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala

```diff
@@ -414,6 +434,296 @@ private[spark] object ClosureCleaner extends Logging {
+    // and all inner classes discovered below.
+    val trackedClassesByInternalName = Map[String, Class[_]](implClassInternalName -> implClass)
+
+    // Depth-first search for inner closures and track the fields that were accessed in them.
+    // Start from the lambda body's implementation method, follow method invocations
+    val visited = Set.empty[MethodIdentifier[_]]
+    val stack = Stack[MethodIdentifier[_]](implMethodId)
+    def pushIfNotVisited(methodId: MethodIdentifier[_]): Unit = {
+      if (!visited.contains(methodId)) {
+        stack.push(methodId)
+      }
+    }
+
+    while (!stack.isEmpty) {
+      val currentId = stack.pop
+      visited += currentId
+
+      val currentClass = currentId.cls
+      val currentMethodNode = methodNodeById(currentId)
+      logTrace(s"  scanning ${currentId.cls.getName}.${currentId.name}${currentId.desc}")
+      currentMethodNode.accept(new MethodVisitor(ASM7) {
+        val currentClassName = currentClass.getName
+        val currentClassInternalName = currentClassName.replace('.', '/')
+
+        // Find and update the accessedFields info. Only fields on known outer classes are tracked.
+        // This is the FieldAccessFinder equivalent.
+        override def visitFieldInsn(op: Int, owner: String, name: String, desc: String): Unit = {
+          if (op == GETFIELD || op == PUTFIELD) {
+            val ownerExternalName = owner.replace('/', '.')
+            for (cl <- accessedFields.keys if cl.getName == ownerExternalName) {
+              logTrace(s"    found field access $name on $ownerExternalName")
+              accessedFields(cl) += name
+            }
+          }
+        }
+
+        override def visitMethodInsn(
+            op: Int, owner: String, name: String, desc: String, itf: Boolean): Unit = {
+          val ownerExternalName = owner.replace('/', '.')
+          if (owner == currentClassInternalName) {
+            logTrace(s"    found intra class call to $ownerExternalName.$name$desc")
+            // could be invoking a helper method or a field accessor method, just follow it.
+            pushIfNotVisited(MethodIdentifier(currentClass, name, desc))
+          } else if (isInnerClassCtorCapturingOuter(
+              op, owner, name, desc, currentClassInternalName)) {
+            // Discover inner classes.
+            // This this the InnerClassFinder equivalent for inner classes, which still use the
+            // `$outer` chain. So this is NOT controlled by the `findTransitively` flag.
+            logTrace(s"    found inner class $ownerExternalName")
+            val innerClassInfo = getOrUpdateClassInfo(owner)
+            val innerClass = innerClassInfo._1
+            val innerClassNode = innerClassInfo._2
+            trackedClassesByInternalName(owner) = innerClass
+            // We need to visit all methods on the inner class so that we don't missing anything.
+            for (m <- innerClassNode.methods.asScala) {
```

Comment:
How about logging found methods here (e.g., `logTrace(s"        found method ${m.name}${m.desc}")`), too?  The additional log looks like this;
```
20/05/09 08:33:01 TRACE IndylambdaScalaClosures:   scanning $line14.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw.$anonfun$closure$1(L$line14/$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw;I)Lscala/collection/immutable/IndexedSeq;
20/05/09 08:33:01 TRACE IndylambdaScalaClosures:     found inner class $line14.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$InnerFoo$1
20/05/09 08:33:01 TRACE IndylambdaScalaClosures:         found method innerClosure()Lscala/Function1;
20/05/09 08:33:01 TRACE IndylambdaScalaClosures:         found method $anonfun$innerClosure$2(L$line14/$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$InnerFoo$1;I)Ljava/lang/String;
20/05/09 08:33:01 TRACE IndylambdaScalaClosures:         found method $anonfun$innerClosure$1(L$line14/$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$InnerFoo$1;I)Lscala/collection/immutable/IndexedSeq;
20/05/09 08:33:01 TRACE IndylambdaScalaClosures:         found method <init>(L$line14/$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw;Ljava/lang/String;)V
20/05/09 08:33:01 TRACE IndylambdaScalaClosures:         found method $anonfun$innerClosure$2$adapted(L$line14/$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$InnerFoo$1;Ljava/lang/Object;)Ljava/lang/String;
20/05/09 08:33:01 TRACE IndylambdaScalaClosures:         found method $anonfun$innerClosure$1$adapted(L$line14/$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$InnerFoo$1;Ljava/lang/Object;)Lscala/collection/immutable/IndexedSeq;
20/05/09 08:33:01 TRACE IndylambdaScalaClosures:         found method $deserializeLambda$(Ljava/lang/invoke/SerializedLambda;)Ljava/lang/Object;
20/05/09 08:33:01 TRACE IndylambdaScalaClosures:     ignoring call to scala.Predef$.intWrapper(I)I
20/05/09 08:33:01 TRACE IndylambdaScalaClosures:     ignoring call to scala.runtime.RichInt$.to$extension0(II)Lscala/collection/immutable/Range$Inclusive;
20/05/09 08:33:01 TRACE IndylambdaScalaClosures:     found call to outer $line14.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$InnerFoo$1.innerClosure()Lscala/Function1;
...
```

Reply:
Thanks for the suggestion! I actually had something very similar to this at the same place for my own debugging. This helped me squash a few bugs in my code that added the wrong class...already fixed before I sent out the PR 🤣 

If we assume that my visiting strategy is correct, then the "scanning ..." log lines will actually contain all the information from these "found method ..." log lines, just at a different timing / place.

It is good to log both the "found"s and "scan"s for consistency check, but I was wondering whether or not that's too verbose even for trace-level logging.

WDYT? I was thinking about maybe going the other way around, which is to promote some of my `logTrace`s to `logDebug`, e.g. the `found inner class` and `found inner closure` ones, since they're present in the debug-level logs in the old implementation. Then I'd be willing to add `logTrace` for "found method".

Reply:
> It is good to log both the "found"s and "scan"s for consistency check

Yea, I think so, too, and the log message looks useful for debugging. Actually, I added this `logTrace` by myself for checking this PR...

Reply:
👍  let me promote the the "found inner class" and "found inner closure" to DEBUG and add the logging you're suggesting here. Thanks!

Reply:
Addressed ^_^